src/OrbitGl/App.cpp

```diff
@@ -1881,6 +1882,135 @@ Future<ErrorMessageOr<CanceledOr<std::filesystem::path>>> OrbitApp::RetrieveModu
   return chained_result_future;
 }
 
+Future<ErrorMessageOr<CanceledOr<std::filesystem::path>>> OrbitApp::RetrieveModuleItself(
+    const ModuleIdentifier& module_id, uint64_t module_file_size) {
+  ORBIT_SCOPE_FUNCTION;
+  ORBIT_CHECK(std::this_thread::get_id() == main_thread_id_);
+
+  auto find_locally_or_in_cache = [this, module_id,
+                                   module_file_size]() -> ErrorMessageOr<std::filesystem::path> {
+    if (absl::GetFlag(FLAGS_local)) {
+      auto verify_object_file_result = orbit_symbols::VerifyObjectFile(
+          module_id.file_path, module_id.build_id, module_file_size);
+      if (verify_object_file_result.has_value()) {
+        ORBIT_LOG("Found module file \"%s\" itself locally", module_id.file_path);
+        return module_id.file_path;
+      }
+      return ErrorMessage{"\n  * Could not find module file itself locally."};
+    }
+
+    auto object_in_cache =
+        symbol_helper_.FindObjectInCache(module_id.file_path, module_id.build_id, module_file_size);
+    if (object_in_cache.has_value()) {
+      ORBIT_LOG("Found module file \"%s\" itself in cache", module_id.file_path);
+      return {object_in_cache.value()};
+    }
+    return ErrorMessage{absl::StrFormat("\n  * Could not find module file itself in cache: %s",
+                                        object_in_cache.error().message())};
+  };
+
+  Future<ErrorMessageOr<std::filesystem::path>> find_locally_or_in_cache_future =
+      thread_pool_->Schedule(find_locally_or_in_cache);
+
+  auto retrieve_from_instance =
+      [this, module_id](
+          const ErrorMessageOr<std::filesystem::path>& find_locally_or_in_cache_result) mutable
+      -> Future<ErrorMessageOr<CanceledOr<std::filesystem::path>>> {
+    if (download_disabled_modules_.contains(module_id.file_path)) {
+      if (find_locally_or_in_cache_result.has_value()) {
+        return {find_locally_or_in_cache_result.value()};
+      }
+      return {find_locally_or_in_cache_result.error()};
+    }
+
+    if (find_locally_or_in_cache_result.has_value()) {
+      return {find_locally_or_in_cache_result.value()};
+    }
```

Comment:
Change the `find_locally_or_in_cache` lambda to return `ErrorMessageOr<CanceledOr<std::filesystem::path>>`
and then
```cpp
Future<ErrorMessageOr<CanceledOr<std::filesystem::path>>> find_locally_or_in_cache_future =
      thread_pool_->Schedule(find_locally_or_in_cache);

if (download_disabled_modules_.contains(module_id.file_path)) return find_locally_or_in_cache_future;

auto retrieve_from_instance =
      [this, module_id](
          const ErrorMessageOr<std::filesystem::path>& find_locally_or_in_cache_result) mutable
      -> Future<ErrorMessageOr<CanceledOr<std::filesystem::path>>> {
    if (find_locally_or_in_cache_result.has_value()) {
      return {find_locally_or_in_cache_result.value()};
    }
...
```